<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Getting up-and-running with sceptre</title>

<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Getting up-and-running with sceptre</h1>



<p>This vignette helps users get up-and-running with the <strong>sceptre</strong> R package. We show how to use <strong>sceptre</strong> to analyze a small subset of a real single-cell pooled CRISPR screen dataset. To download <strong>sceptre</strong>, follow the instructions in the <strong>sceptre</strong> <a href="https://github.com/Timothy-Barry/sceptre_paper">Github repository</a>.</p>
<p>We consider data from the paper <a href="https://www.cell.com/cell/fulltext/S0092-8674(18)31554-X?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS009286741831554X%3Fshowall%3Dtrue">A Genome-wide Framework for Mapping Gene Regulation via Cellular Genetic Screens</a> published by Gasperini et al. in Cell in 2019. The authors perturbed candidate enhancers using CRISPRi and assessed the impact of these perturbations on gene expression through sc-RNAseq. Our goal in this section is to determine the impact of a single perturbation on a single gene, accounting for appropriate technical variables. The perturbation and gene that we investigate were randomly selected from the set of all perturbations and genes in the Gasperini et al. dataset. We begin by loading the data.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">library</span>(sceptre)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="kw">data</span>(<span class="st">&quot;example_data&quot;</span>)</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="kw">head</span>(example_data)</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="co">#&gt; # A tibble: 6 x 7</span></span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="co">#&gt;   gene_exp gRNA_indic  p_mito prep_batch total_umis guide_count gene_count</span></span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="co">#&gt;      &lt;int&gt;      &lt;dbl&gt;   &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;       &lt;dbl&gt;      &lt;int&gt;</span></span>
<span id="cb1-7"><a href="#cb1-7"></a><span class="co">#&gt; 1        2          0 0.0588           1      17572          67       3549</span></span>
<span id="cb1-8"><a href="#cb1-8"></a><span class="co">#&gt; 2        1          0 0.0361           1       8923          26       2543</span></span>
<span id="cb1-9"><a href="#cb1-9"></a><span class="co">#&gt; 3        0          0 0.0698           1      14637          61       3191</span></span>
<span id="cb1-10"><a href="#cb1-10"></a><span class="co">#&gt; 4        1          0 0.0262           1      22798          39       4539</span></span>
<span id="cb1-11"><a href="#cb1-11"></a><span class="co">#&gt; 5        0          0 0.00799          1      10136          37       2605</span></span>
<span id="cb1-12"><a href="#cb1-12"></a><span class="co">#&gt; 6        1          0 0.0224           1       9751          57       2187</span></span></code></pre></div>
<p>The <em>example_data</em> data frame contains 205,797 rows and 7 columns. Rows correspond to cells, and columns correspond to cell-specific features. The first column records gene expression (in UMIs), the second records gRNA presence or absence, and the final five record additional cell-specific covariates (fraction of UMIs that map to mitochondrial genes, batch effect, total UMI count across all genes, total gRNA count across all guides, and total genes expressed). We define the vectors <em>expressions</em> and <em>gRNA_indicators</em> and data frame <em>covariate_matrix</em> that we will pass as arguments to the <strong>sceptre</strong> algorithm.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>expressions &lt;-<span class="st"> </span>example_data<span class="op">$</span>gene_exp</span>
<span id="cb2-2"><a href="#cb2-2"></a>expressions[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="co">#&gt; [1] 2 1 0 1 0</span></span>
<span id="cb2-4"><a href="#cb2-4"></a></span>
<span id="cb2-5"><a href="#cb2-5"></a>gRNA_indicators &lt;-<span class="st"> </span>example_data<span class="op">$</span>gRNA_indic</span>
<span id="cb2-6"><a href="#cb2-6"></a>gRNA_indicators[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span>
<span id="cb2-7"><a href="#cb2-7"></a><span class="co">#&gt; [1] 0 0 0 0 0</span></span>
<span id="cb2-8"><a href="#cb2-8"></a></span>
<span id="cb2-9"><a href="#cb2-9"></a>covariate_matrix &lt;-<span class="st"> </span><span class="kw">summarize</span>(example_data, </span>
<span id="cb2-10"><a href="#cb2-10"></a>                              <span class="dt">p_mito =</span> p_mito, </span>
<span id="cb2-11"><a href="#cb2-11"></a>                              <span class="dt">prep_batch =</span> <span class="kw">factor</span>(prep_batch), </span>
<span id="cb2-12"><a href="#cb2-12"></a>                              <span class="dt">lg_total_umis =</span> <span class="kw">log</span>(total_umis), </span>
<span id="cb2-13"><a href="#cb2-13"></a>                              <span class="dt">lg_guide_count =</span> <span class="kw">log</span>(guide_count), </span>
<span id="cb2-14"><a href="#cb2-14"></a>                              <span class="dt">lg_gene_count =</span> <span class="kw">log</span>(gene_count))</span>
<span id="cb2-15"><a href="#cb2-15"></a><span class="kw">head</span>(covariate_matrix)</span>
<span id="cb2-16"><a href="#cb2-16"></a><span class="co">#&gt; # A tibble: 6 x 5</span></span>
<span id="cb2-17"><a href="#cb2-17"></a><span class="co">#&gt;    p_mito prep_batch lg_total_umis lg_guide_count lg_gene_count</span></span>
<span id="cb2-18"><a href="#cb2-18"></a><span class="co">#&gt;     &lt;dbl&gt; &lt;fct&gt;              &lt;dbl&gt;          &lt;dbl&gt;         &lt;dbl&gt;</span></span>
<span id="cb2-19"><a href="#cb2-19"></a><span class="co">#&gt; 1 0.0588  1                   9.77           4.20          8.17</span></span>
<span id="cb2-20"><a href="#cb2-20"></a><span class="co">#&gt; 2 0.0361  1                   9.10           3.26          7.84</span></span>
<span id="cb2-21"><a href="#cb2-21"></a><span class="co">#&gt; 3 0.0698  1                   9.59           4.11          8.07</span></span>
<span id="cb2-22"><a href="#cb2-22"></a><span class="co">#&gt; 4 0.0262  1                  10.0            3.66          8.42</span></span>
<span id="cb2-23"><a href="#cb2-23"></a><span class="co">#&gt; 5 0.00799 1                   9.22           3.61          7.87</span></span>
<span id="cb2-24"><a href="#cb2-24"></a><span class="co">#&gt; 6 0.0224  1                   9.19           4.04          7.69</span></span></code></pre></div>
<p>The covariate matrix consists of all technical variables that will be accounted for using <strong>sceptre</strong>. Users of <strong>sceptre</strong> need to decide which variables to include in the covariate matrix given their analysis objective and data; reasonable choices include the fraction of UMIs that map to mitochondrial genes, batch effect, and log-transformed total UMI count (across all genes). The <em>expressions</em> vector is untransformed and unscaled; <strong>sceptre</strong> expects gene expressions to be provided as raw UMI counts.</p>
<p>Next, we run the sceptre algorithm by calling the core function <em>run_sceptre_gRNA_gene_pair</em>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>p_value &lt;-<span class="st"> </span><span class="kw">run_sceptre_gRNA_gene_pair</span>(<span class="dt">expressions =</span> expressions, </span>
<span id="cb3-2"><a href="#cb3-2"></a>                                      <span class="dt">gRNA_indicators =</span> gRNA_indicators, </span>
<span id="cb3-3"><a href="#cb3-3"></a>                                      <span class="dt">covariate_matrix =</span> covariate_matrix, </span>
<span id="cb3-4"><a href="#cb3-4"></a>                                      <span class="dt">seed =</span> <span class="dv">1234</span>)</span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="co">#&gt; Running resample 100/500</span></span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="co">#&gt; Running resample 200/500</span></span>
<span id="cb3-7"><a href="#cb3-7"></a><span class="co">#&gt; Running resample 300/500</span></span>
<span id="cb3-8"><a href="#cb3-8"></a><span class="co">#&gt; Running resample 400/500</span></span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="co">#&gt; Running resample 500/500</span></span>
<span id="cb3-10"><a href="#cb3-10"></a>p_value</span>
<span id="cb3-11"><a href="#cb3-11"></a><span class="co">#&gt; [1] 0.2610861</span></span></code></pre></div>
<p>The output, <em>p_value</em>, is a numeric scalar containing the p-value of a test of the null hypothesis that the perturbation has no effect on gene expression. In our example the p-value is fairly large (p = 0.261), indicating that there is no evidence that the perturbation affects gene expression.</p>
<p>The <em>run_sceptre_gRNA_gene_pair</em> function has several optional parameters, including <em>gRNA_precomp</em>, <em>gene_precomp_dispersion</em>, and <em>gene_precomp_offsets</em> (see function documentation and <strong>sceptre</strong> <a href="https://www.biorxiv.org/content/10.1101/2020.08.13.250092v2">manuscript</a> for details). The user can pass values to these parameters to accelerate the execution of the function, a good idea when analyzing a large number of gRNA-gene pairs. If the user chooses not to pass values to these optional parameters, <em>run_sceptre_gRNA_gene_pair</em> calculates these quantities in a subroutine. The other optional parameters of <em>run_sceptre_gRNA_gene_pair</em> are <em>B</em> and <em>seed</em>. <em>B</em> is the number of resamples to draw in the conditional randomization test subroutine (default 500), and <em>seed</em> is the seed passed to the random number generator (default none).</p>
<p>Currently, <strong>sceptre</strong> can be used to test for genomic elements that up-regulate gene expression (e.g., enhancers, promoters). <strong>sceptre</strong> does not support testing for genomic elements that down-regulate gene expression (e.g., silencers); this functionality likely will be added to a future iteration of the package.</p>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
