---
title: "Getting up-and-running with sceptre"
authors: Tim Barry, Gene Katsevich, Kathryn Roeder
output: html_vignette
vignette: >
  %\VignetteIndexEntry{Getting up-and-running with sceptre}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This vignette helps users get up-and-running with the **sceptre** R package. We show how to use **sceptre** to analyze a small subset of a real single-cell pooled CRISPR screen dataset. To download **sceptre**, follow the instructions in the **sceptre** [Github repository](https://github.com/Timothy-Barry/sceptre_paper).

We consider data from the paper [A Genome-wide Framework for Mapping Gene Regulation via Cellular Genetic Screens](https://www.cell.com/cell/fulltext/S0092-8674(18)31554-X?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS009286741831554X%3Fshowall%3Dtrue) published by Gasperini et al. in Cell in 2019. The authors perturbed candidate enhancers using CRISPRi and assessed the impact of these perturbations on gene expression through sc-RNAseq. Our goal in this section is to determine the impact of a single perturbation on a single gene, accounting for appropriate technical variables. The perturbation and gene that we investigate were randomly selected from the set of all perturbations and genes in the Gasperini et al. dataset. We begin by loading the data.

```{r}
library(sceptre)
data("example_data")
head(example_data)
```
The *example_data* data frame contains 205,797 rows and 7 columns. Rows correspond to cells, and columns correspond to cell-specific features. The first column records gene expression (in UMIs), the second records gRNA presence or absence, and the final five record additional cell-specific covariates (fraction of UMIs that map to mitochondrial genes, batch effect, total UMI count across all genes, total gRNA count across all guides, and total genes expressed). We define the vectors *expressions* and *gRNA_indicators* and data frame *covariate_matrix* that we will pass as arguments to the **sceptre** algorithm.

```{r}
expressions <- example_data$gene_exp
expressions[1:5]

gRNA_indicators <- example_data$gRNA_indic
gRNA_indicators[1:5]

covariate_matrix <- summarize(example_data, 
                              p_mito = p_mito, 
                              prep_batch = factor(prep_batch), 
                              lg_total_umis = log(total_umis), 
                              lg_guide_count = log(guide_count), 
                              lg_gene_count = log(gene_count))
head(covariate_matrix)
```

The covariate matrix consists of all technical variables that will be accounted for using **sceptre**. Users of **sceptre** need to decide which variables to include in the covariate matrix given their analysis objective and data; reasonable choices include the fraction of UMIs that map to mitochondrial genes, batch effect, and log-transformed total UMI count (across all genes). The *expressions* vector is untransformed and unscaled; **sceptre** expects gene expressions to be provided as raw UMI counts.

Next, we run the sceptre algorithm by calling the core function *run_sceptre_gRNA_gene_pair*.

```{r}
p_value <- run_sceptre_gRNA_gene_pair(expressions = expressions, 
                                      gRNA_indicators = gRNA_indicators, 
                                      covariate_matrix = covariate_matrix, 
                                      seed = 1234)
p_value
```

The output, *p_value*, is a numeric scalar containing the p-value of a test of the null hypothesis that the perturbation has no effect on gene expression. In our example the p-value is fairly large (p = 0.261), indicating that there is no evidence that the perturbation affects gene expression.

The *run_sceptre_gRNA_gene_pair* function has several optional parameters, including *gRNA_precomp*, *gene_precomp_dispersion*, and *gene_precomp_offsets* (see function documentation and **sceptre** [manuscript](https://www.biorxiv.org/content/10.1101/2020.08.13.250092v2) for details). The user can pass values to these parameters to accelerate the execution of the function, a good idea when analyzing a large number of gRNA-gene pairs. If the user chooses not to pass values to these optional parameters, *run_sceptre_gRNA_gene_pair* calculates these quantities in a subroutine. The other optional parameters of *run_sceptre_gRNA_gene_pair* are *B* and *seed*. *B* is the number of resamples to draw in the conditional randomization test subroutine (default 500), and *seed* is the seed passed to the random number generator (default none).

Currently, **sceptre** can be used to test for genomic elements that up-regulate gene expression (e.g., enhancers, promoters). **sceptre** does not support testing for genomic elements that down-regulate gene expression (e.g., silencers); this functionality likely will be added to a future iteration of the package.
